Build Fails If Cabal is Installed

# Steps to Reproduce

1. Install Cabal via Homebrew

   ```sh
   brew install cabal-install
   ```

2. Create the build system

   ```sh
   mkdir -p build
   cd build; cmake ..                                 \
         -GNinja                                      \
         -DENABLE_ASAN=OFF                            \
         -DENABLE_LOGGER=ON                           \
         -DENABLE_DEBUG=ON                            \
         -DCMAKE_EXPORT_COMPILE_COMMANDS=ON           \
         -DBUILD_DOCUMENTATION=OFF                    \
         -DBUILD_DOCSET=OFF                           \
         -DBUILD_PDF=OFF                              \
         -DPDFLATEX_COMPILER=`which latexmk`          \
         -DPDFLATEX_COMPILER_OPTIONS='-pdf;-f;-quiet' \
         -DCMAKE_PREFIX_PATH=/usr/local/opt/qt5       \
         -DTOOLS="ALL"                                \
         -DBINDINGS="ALL;-swig_python2;-swig_ruby"    \
         -DPLUGINS="ALL;-jni;-python2;-lua"
   ```

# Expected Result

CMake configures the project without any problems.

# Actual Result

CMake prints the following error message:

```
CMake Error at cmake/Modules/LibAddPlugin.cmake:393 (add_dependencies):
  The dependency target "c2hs_haskell" of target "elektra-haskell" does not
  exist.
Call Stack (most recent call first):
  cmake/Modules/LibAddHaskellPlugin.cmake:141 (add_plugin)
  src/plugins/haskell/CMakeLists.txt:3 (add_haskell_plugin)


CMake Error at cmake/Modules/LibAddPlugin.cmake:353 (add_dependencies):
  The dependency target "c2hs_haskell" of target "elektra-haskell-objects"
  does not exist.
Call Stack (most recent call first):
  cmake/Modules/LibAddHaskellPlugin.cmake:141 (add_plugin)
  src/plugins/haskell/CMakeLists.txt:3 (add_haskell_plugin)
```

.

# System Information

- Elektra Version: master
- OS: macOS 10.13

I get the error:

```
 CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
 Please set them or make sure they are set and tested correctly in the CMake files:
 GHC_FFI_LIB
     linked by target "elektra-haskell" in directory /home/markus/Projekte/Elektra/current/src/plugins/haskell
     linked by target "testmod_haskell" in directory /home/markus/Projekte/Elektra/current/src/plugins
 GHC_RTS_LIB
     linked by target "elektra-haskell" in directory /home/markus/Projekte/Elektra/current/src/plugins/haskell
     linked by target "testmod_haskell" in directory /home/markus/Projekte/Elektra/current/src/plugins
```

And we should tag all the GHC_ cmake variables as INTERNAL.

--- a/cmake/Modules/LibAddHaskellPlugin.cmake
+++ b/cmake/Modules/LibAddHaskellPlugin.cmake
@@ -150,7 +150,7 @@ macro (add_haskell_plugin target)
 			endif (GHC_FFI_LIB)
 
 		else (FINDEX GREATER -1)
-			remove_plugin (${target} "haskell bindings are not included in the make configuration")
+			remove_plugin (${target} "haskell bindings are not included in the cmake configuration")
 		endif (FINDEX GREATER -1)
 		else (GHC-PKG_EXECUTABLE)
 			remove_plugin (${target} "ghc-pkg not found")

--- a/cmake/Modules/LibAddHaskellPlugin.cmake
+++ b/cmake/Modules/LibAddHaskellPlugin.cmake
@@ -130,6 +130,7 @@ macro (add_haskell_plugin target)
 				COMMAND ${CABAL_EXECUTABLE} --ipid=${target} configure
 				COMMAND ${CABAL_EXECUTABLE} build
 				WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+				BYPRODUCTS ${CMAKE_CURRENT_BINARY_DIR}/dist/build/libHS${target}.a
 				DEPENDS ${target}-register
 			)
 

--- a/cmake/Modules/LibAddHaskellPlugin.cmake
+++ b/cmake/Modules/LibAddHaskellPlugin.cmake
@@ -17,8 +17,8 @@ macro (add_haskell_plugin target)
 		${ARGN}
 	)
 
-	SET (PLUGIN_NAME ${target})
-	SET (PLUGIN_NAME_CAPITALIZED ${target})
+	set (PLUGIN_NAME ${target})
+	set (PLUGIN_NAME_CAPITALIZED ${target})
 	string (SUBSTRING ${PLUGIN_NAME} 0 1 FIRST_LETTER)
 	string (TOUPPER ${FIRST_LETTER} FIRST_LETTER)
 	string (REGEX REPLACE "^.(.*)" "${FIRST_LETTER}\\1" PLUGIN_NAME_CAPITALIZED "${PLUGIN_NAME}")
@@ -33,6 +33,8 @@ macro (add_haskell_plugin target)
 		if (CABAL_EXECUTABLE)
 		if (GHC_EXECUTABLE)
 		if (GHC-PKG_EXECUTABLE)
+		list (FIND BINDINGS "haskell" FINDEX)
+		if (FINDEX GREATER -1)
 
 			# needed for HsFFI.h
 			execute_process (
@@ -71,6 +73,12 @@ macro (add_haskell_plugin target)
 				OUTPUT_VARIABLE GHC_PRIM_NAME OUTPUT_STRIP_TRAILING_WHITESPACE
 			)
 			find_library (GHC_PRIM_LIB "HS${GHC_PRIM_NAME}" ${GHC_LIB_DIR}/${GHC_PRIM_NAME})
+
+			if (GHC_FFI_LIB)
+			if (GHC_RTS_LIB)
+			if (GHC_BASE_LIB)
+			if (GHC_GMP_LIB)
+			if (GHC_PRIM_LIB)
 
 			set (GHC_LIB_DIRS
 				"${CMAKE_CURRENT_BINARY_DIR}/dist/build/libHS${target}.a"
@@ -125,6 +133,25 @@ macro (add_haskell_plugin target)
 				DEPENDS ${target}-register
 			)
 
+			else (GHC_PRIM_LIB)
+				remove_plugin (${target} "GHC_PRIM_LIB not found")
+			endif (GHC_PRIM_LIB)
+			else (GHC_GMP_LIB)
+				remove_plugin (${target} "GHC_GMP_LIB not found")
+			endif (GHC_GMP_LIB)
+			else (GHC_BASE_LIB)
+				remove_plugin (${target} "GHC_BASE_LIB not found")
+			endif (GHC_BASE_LIB)
+			else (GHC_RTS_LIB)
+				remove_plugin (${target} "GHC_RTS_LIB not found")
+			endif (GHC_RTS_LIB)
+			else (GHC_FFI_LIB)
+				remove_plugin (${target} "GHC_FFI_LIB not found")
+			endif (GHC_FFI_LIB)
+
+		else (FINDEX GREATER -1)
+			remove_plugin (${target} "haskell bindings are not included in the make configuration")
+		endif (FINDEX GREATER -1)
 		else (GHC-PKG_EXECUTABLE)
 			remove_plugin (${target} "ghc-pkg not found")
 		endif (GHC-PKG_EXECUTABLE)
@@ -149,5 +176,16 @@ macro (add_haskell_plugin target)
 		DEPENDS
 			${target} c2hs_haskell
 		ADD_TEST
+	)
+
+	mark_as_advanced (
+		GHC_EXECUTABLE
+		GHC-PKG_EXECUTABLE
+		CABAL_EXECUTABLE
+		GHC_FFI_LIB
+		GHC_RTS_LIB
+		GHC_BASE_LIB
+		GHC_GMP_LIB
+		GHC_PRIM_LIB
 	)
 endmacro (add_haskell_plugin)

--- a/src/bindings/haskell/CMakeLists.txt
+++ b/src/bindings/haskell/CMakeLists.txt
@@ -19,6 +19,7 @@ if (GHC_EXECUTABLE)
 		c2hs_haskell ALL
 		COMMAND ${CABAL_EXECUTABLE} --ipid=libelektra-haskell-${KDB_VERSION} configure
 		COMMAND ${CABAL_EXECUTABLE} build
+		BYPRODUCTS ${CMAKE_CURRENT_BINARY_DIR}/dist/build/libHSlibelektra-haskell-${KDB_VERSION}.a
 		WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
 	)
 

--- a/src/bindings/haskell/CMakeLists.txt
+++ b/src/bindings/haskell/CMakeLists.txt
@@ -6,8 +6,7 @@ if (CABAL_EXECUTABLE) # set by find_program
 if (C2HS_EXECUTABLE)
 if (GHC_EXECUTABLE)
 	
-	SET (CABAL_INCLUDE_DIRS "${CMAKE_SOURCE_DIR}/src/include, ${CMAKE_BINARY_DIR}/src/include")
-	SET (CABAL_EXTRA_LIB_DIRS "")
+	set (CABAL_INCLUDE_DIRS "${CMAKE_SOURCE_DIR}/src/include, ${CMAKE_BINARY_DIR}/src/include")
 
 	# configure include paths
 	configure_file (
@@ -54,3 +53,9 @@ endif (C2HS_EXECUTABLE)
 else (CABAL_EXECUTABLE)
 	remove_binding (c2hs_haskell "cabal not found")
 endif (CABAL_EXECUTABLE)
+
+mark_as_advanced (
+	GHC_EXECUTABLE
+	C2HS_EXECUTABLE
+	CABAL_EXECUTABLE
+)

--- a/src/bindings/haskell/src/Elektra/Key.chs
+++ b/src/bindings/haskell/src/Elektra/Key.chs
@@ -122,7 +122,7 @@ keyListMeta key = keyRewindMeta key >> listMeta []
         listMeta res = do
             cur <- keyNextMeta key
             isNull <- keyPtrNull cur
-            if isNull then return res else liftM (cur :) (listMeta res)
+            if isNull then return res else keyIncRef cur >> liftM (cur :) (listMeta res)
 
 -- ***
 -- KEY TESTING METHODS

